20230911 doc 8558 add zone config extension example

rmloveland opened the pull request.

rmloveland commented:
@dikshant I guess zone config extensions owned by foundations now? is there someone on the eng side you would recommend have a look at this as well?

re: the PR content, the super regions example is a pretty straight easy port to `cockroach demo` with a multi-region movr DB

the x-region write latency I had to tweak a bit b/c once movr becomes multi-region all tables have REGIONAL BY TABLE already which is nice.  I had to remove the setting of `lease_preferences` that was given in the Jira issue b/c that is [not allowed by zone config extensions](http://cockroachlabs.com/docs/v23.2/zone-config-extensions.html), but presumably making the distant regions have non-voting replicas is enough?

ZhouXing19 reviewed (approved):
LGTM, thanks!

ZhouXing19 commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+{% include_cached copy-clipboard.html %}
+~~~ sql
+ALTER DATABASE movr SET PRIMARY REGION "us-east1";
+ALTER DATABASE movr SURVIVE REGION FAILURE;
```
The `ALTER DATABASE movr SURVIVE REGION FAILURE;` will need to go after the `ADD REGION` commands.

rmloveland replied:
Thanks, this is updated in the latest commit

ZhouXing19 commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+{% include_cached copy-clipboard.html %}
+~~~ sql
+ALTER DATABASE movr ALTER LOCALITY REGIONAL IN "us-east1" CONFIGURE ZONE USING
+    num_replicas = 5,
```
We have 6 regions for the cluster, so with goal `REGION_FAILURE`, the num_replicas should be at least `numReplicas = maxFailuresBeforeUnavailability(numVotersForRegionSurvival) + (numRegions - 1)`, which is 2 + 6 - 1 = 7 in our case. I think we can do 
```
ALTER DATABASE movr ALTER LOCALITY REGIONAL IN "us-east1" CONFIGURE ZONE USING
    num_replicas = 7,
    constraints = '{+region=us-east1: 3,
                    +region=us-central1: 3,
                    +region=us-west1: 1}',
    lease_preferences = '[[+region=us-east1], [+region=us-central1]]';
```

rmloveland replied:
Thanks Jane, this is updated in the latest commit.  

I updated both the "us-east1" locality and the "europe-central1" locality with the `num_replicas=7` and the constraints to be `3,3,1`

ZhouXing19 commented:
> the x-region write latency I had to tweak a bit b/c once movr becomes multi-region all tables have REGIONAL BY TABLE already which is nice. I had to remove the setting of lease_preferences that was given in the Jira issue b/c that is [not allowed by zone config extensions](http://cockroachlabs.com/docs/v23.2/zone-config-extensions.html), but presumably making the distant regions have non-voting replicas is enough?

Good question, I'm not sure for the `REGIONAL` case, though it's not allowed to set `lease_preference` with zone config extension, whether we should set it with `ALTER DATABASE movr CONFIGURE ZONE USING lease_preference` prior to setting the zone config extension. Started a thread in https://cockroachlabs.slack.com/archives/C4ERHM60Z/p1696266134754579

rmloveland commented:
> > the x-region write latency I had to tweak a bit b/c once movr becomes multi-region all tables have REGIONAL BY TABLE already which is nice. I had to remove the setting of lease_preferences that was given in the Jira issue b/c that is [not allowed by zone config extensions](http://cockroachlabs.com/docs/v23.2/zone-config-extensions.html), but presumably making the distant regions have non-voting replicas is enough?
> 
> Good question, I'm not sure for the `REGIONAL` case, though it's not allowed to set `lease_preference` with zone config extension, whether we should set it with `ALTER DATABASE movr CONFIGURE ZONE USING lease_preference` prior to setting the zone config extension. Started a thread in https://cockroachlabs.slack.com/archives/C4ERHM60Z/p1696266134754579

ok I'll jump on that thread! thanks

rmloveland commented:
@ecwall and @dikshant , removing y'all since Jane has LGTM'd (Thanks Jane!!!)

Gonna send this on for docs review so we can get it merged

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
 
+#### Implement super regions
+
+In this example, we use [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) to provide an alternative implementation of [Super Regions]({% link {{ page.version.version }}/multiregion-overview.md %}#super-regions), which are useful for [data domiciling]({% link {{ page.version.version }}/data-domiciling.md %}).
```
Per style (feel free to replace my suggestion though) https://github.com/cockroachdb/docs/blob/main/StyleGuide.md#style-and-tone

```suggestion
In this example, [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) are used to provide an alternative implementation of [Super Regions]({% link {{ page.version.version }}/multiregion-overview.md %}#super-regions), which are useful for [data domiciling]({% link {{ page.version.version }}/data-domiciling.md %}).
```

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+
+#### Minimize cross-region write latency
+
+In this example, we use [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) to minimize cross-region write latency in a cluster with the following characteristics:
```
```suggestion
In this example, [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) are used to minimize cross-region write latency in a cluster with the following characteristics:
```

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+
+In this example, we use [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) to minimize cross-region write latency in a cluster with the following characteristics:
+
+- 3 main US regions (`us-west1`,`us-east1`,`us-central1`).
```
Seems best to spell out a numeral that begins a sentence, but it's just a guess

```suggestion
- Three main US regions (`us-west1`,`us-east1`,`us-central1`).
```

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+In this example, we use [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) to minimize cross-region write latency in a cluster with the following characteristics:
+
+- 3 main US regions (`us-west1`,`us-east1`,`us-central1`).
+- 2 additional distant regions (`europe-west1`, `asia-northeast1`). Only [follower reads]({% link {{ page.version.version }}/follower-reads.md %}) will be made from these regions.
```
```suggestion
- Two additional distant regions (`europe-west1`, `asia-northeast1`). Only [follower reads]({% link {{ page.version.version }}/follower-reads.md %}) will be made from these regions.
```

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+- 3 main US regions (`us-west1`,`us-east1`,`us-central1`).
+- 2 additional distant regions (`europe-west1`, `asia-northeast1`). Only [follower reads]({% link {{ page.version.version }}/follower-reads.md %}) will be made from these regions.
+
+    To minimize the cross-region write latency, we will [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that the 3 US regions have voting replicas (that is, replicas that participate in [Raft quorum]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft)). This will leave the 2 additional non-US regions with only [non ...
```
```suggestion
To minimize the cross-region write latency, [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that the 3 US regions have voting replicas (that is, replicas that participate in [Raft quorum]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft)). This will leave the 2 additional non-US regions with only [non-voting replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#non-voting-replicas).
```

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+cockroach demo --nodes=5 --demo-locality=region=us-east1:region=us-west1:region=us-central1:region=europe-west1:region=asia-northeast1
+~~~
+
+Next, enter the statements below at the SQL prompt. These statements will:
```
```suggestion
Next, enter the following statements at the SQL prompt. These statements will:
```

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+
+Next, enter the statements below at the SQL prompt. These statements will:
+
+- Make the [`movr` database]({% link {{ page.version.version }}/movr.md %}) into a [multiregion database]({% link {{ page.version.version }}/multiregion-overview.md %}) by [setting the primary region]({% link {{ page.version.version }}/alter-database.md %}#set-primary-region).
```
```suggestion
- Make the [`movr` database]({% link {{ page.version.version }}/movr.md %}) into a [multi-region database]({% link {{ page.version.version }}/multiregion-overview.md %}) by [setting the primary region]({% link {{ page.version.version }}/alter-database.md %}#set-primary-region).
```

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+
+- Make the [`movr` database]({% link {{ page.version.version }}/movr.md %}) into a [multiregion database]({% link {{ page.version.version }}/multiregion-overview.md %}) by [setting the primary region]({% link {{ page.version.version }}/alter-database.md %}#set-primary-region).
+- Set the `movr` database to [survive region failures]({% link {{ page.version.version }}/multiregion-survival-goals.md %}#survive-region-failures).
+- Add the remaining regions we defined at cluster start time to the multi-region `movr` database (for a total of 5).
```
```suggestion
- Add the remaining regions defined at cluster start time to the multi-region `movr` database (for a total of 5).
```

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+(6 rows)
+~~~
+
+Now that our multiregion cluster is configured, we will use the `ALTER LOCALITY` statements shown below to [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that:
```
Link ALTER LOCALITY. 

```suggestion
Now that the multi-region cluster is configured, use the `ALTER LOCALITY` statements shown below to [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that:
```

taroface commented on src/current/v23.2/alter-database.md:
```diff
@@ -834,6 +836,120 @@ The `lease_preferences` field is now updated to include `us-west1`.
+Now that our multiregion cluster is configured, we will use the `ALTER LOCALITY` statements shown below to [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that:
+
+- The 3 US regions have voting replicas (that is, replicas that participate in [Raft quorum]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft)).
+- This will leave the 2 additional regions (Europe and Japan) with only [non-voting replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#non-voting-replicas) that can be used to serve [follower reads]({% link {{ page.version.version }}/follower-reads.md %}).
```
```suggestion
- The 2 additional regions (Europe and Japan) are left with only [non-voting replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#non-voting-replicas) that can be used to serve [follower reads]({% link {{ page.version.version }}/follower-reads.md %}).
```

taroface commented on src/current/v23.2/cockroach-demo.md:
```diff
@@ -112,7 +112,7 @@ Flag | Description
 `--auto-enable-rangefeeds` | Override the default behavior of `cockroach demo`, which has rangefeeds enabled on startup.  If you do not need to use [changefeeds]({% link {{ page.version.version }}/create-and-configure-changefeeds.md %}) with your demo cluster, use `--auto-enable-rangefeeds=false` to disable rangefeeds and improve performance. See [Enable rangefeeds]({% link {{ page.version.versio ...
 `--cache` | For each demo node, the total size for caches. This can be a percentage (notated as a decimal or with `%`) or any bytes-based unit, for example: <br><br>`--cache=.25`<br>`--cache=25%`<br>`--cache=1000000000 ----> 1000000000 bytes`<br>`--cache=1GB ----> 1000000000 bytes`<br>`--cache=1GiB ----> 1073741824 bytes` <br><br>**Default:** `64MiB`
-`--demo-locality` | Specify [locality]({% link {{ page.version.version }}/cockroach-start.md %}#locality) information for each demo node. The input is a colon-separated list of key-value pairs, where the i<sup>th</sup> pair is the locality setting for the i<sup>th</sup> demo cockroach node.<br><br>For example, the following option assigns node 1's region to `us-east1` and availability zone to `1` ...
+`--demo-locality` | <a name="demo-locality"></a> Specify [locality]({% link {{ page.version.version }}/cockroach-start.md %}#locality) information for each demo node. The input is a colon-separated list of key-value pairs, where the i<sup>th</sup> pair is the locality setting for the i<sup>th</sup> demo cockroach node.<br><br>For example, the following option assigns node 1's region to `us-east1` ...
```
```suggestion
`--demo-locality` | <a name="demo-locality"></a> Specify [locality]({% link {{ page.version.version }}/cockroach-start.md %}#locality) information for each demo node. The input is a colon-separated list of key-value pairs, where the <i>i</i><sup>th</sup> pair is the locality setting for the <i>i</i><sup>th</sup> demo cockroach node.<br><br>For example, the following option assigns node 1's region to `us-east1` and availability zone to `1`, node 2's region to `us-east2` and availability zone to `2`, and node 3's region to `us-east3` and availability zone to `3`:<br><br>`--demo-locality=region=us-east1,az=1:region=us-east1,az=2:region=us-east1,az=3`<br><br>By default, `cockroach demo` uses sample region (`region`) and availability zone (`az`) replica localities for each node specified with the `--nodes` flag.
```

taroface reviewed (approved):
Looks good, just style and formatting nits!

rmloveland commented:
> Looks good, just style and formatting nits!

Awesome, thanks Ryan!  I'm also gonna port all these changes back to
at least v23.1

(Update: patches failed to apply to v22.2 so skipping for now)

> In src/current/v23.2/alter-database.md:
>
> Looks like this is lowercase?

Sure why not?

> ⬇️ Suggested change 
> -- How you could [implement Super regions](#implement-super-regions)
> +- How you could [implement super regions](#implement-super-regions).

Check

> ⬇️ Suggested change 
> -- How to [minimize cross-region write latency](#minimize-cross-region-write-latency)
> +- How to [minimize cross-region write latency](#minimize-cross-region-write-latency).

Fixed

> ⬇️ Suggested change 
> -In this example, we use [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) to provide an alternative implementation of [Super Regions]({% link {{ page.version.version }}/multiregion-overview.md %}#super-regions), which are useful for [data domiciling]({% link {{ page.version.version }}/data-domiciling.md %}).
> +In this example, we use [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) to provide an alternative implementation of [super regions]({% link {{ page.version.version }}/multiregion-overview.md %}#super-regions), which are useful for [data domiciling]({% link {{ page.version.version }}/data-domiciling.md %}).

Fixed

> ⬇️ Suggested change 
> -Next, enter the statements below at the SQL prompt. These statements will:
> +Next, enter the following statements at the SQL prompt. These statements will:

Fixed

> ⬇️ Suggested change 
> -- Make the [`movr` database]({% link {{ page.version.version }}/movr.md %}) into a [multiregion database]({% link {{ page.version.version }}/multiregion-overview.md %}) by [setting the primary region]({% link {{ page.version.version }}/alter-database.md %}#set-primary-region).
> +- Make the [`movr` database]({% link {{ page.version.version }}/movr.md %}) into a [multi-region database]({% link {{ page.version.version }}/multiregion-overview.md %}) by [setting the primary region]({% link {{ page.version.version }}/alter-database.md %}#set-primary-region).

Done. But I'm never going to stop trying to make "fetch" happen!  (fetch = multiregion)

> ⬇️ Suggested change 
> -Now that our multiregion cluster is configured, we will use the `ALTER LOCALITY` statements shown below to [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that:
> +Now that the multi-region cluster is configured, use the following `ALTER LOCALITY` statements to [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that:

Done

> ⬇️ Suggested change 
> -- Add the remaining regions we defined at cluster start time to the multi-region `movr` database (for a total of 6).
> +- Add the remaining regions defined at cluster start time to the multi-region `movr` database (for a total of 6).

Done

> Per style (feel free to replace my suggestion though)
> https://github.com/cockroachdb/docs/blob/main/StyleGuide.md#style-and-tone
>
>
> ⬇️ Suggested change 
> -In this example, we use [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) to provide an alternative implementation of [Super Regions]({% link {{ page.version.version }}/multiregion-overview.md %}#super-regions), which are useful for [data domiciling]({% link {{ page.version.version }}/data-domiciling.md %}).
> +In this example, [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) are used to provide an alternative implementation of [Super Regions]({% link {{ page.version.version }}/multiregion-overview.md %}#super-regions), which are useful for [data domiciling]({% link {{ page.version.version }}/data-domiciling.md %}).

Done

> ⬇️ Suggested change 
> -In this example, we use [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) to minimize cross-region write latency in a cluster with the following characteristics:
> +In this example, [Zone Config Extensions]({% link {{ page.version.version }}/zone-config-extensions.md %}) are used to minimize cross-region write latency in a cluster with the following characteristics:

Done

> ⬇️ Suggested change 
> -- 3 main US regions (`us-west1`,`us-east1`,`us-central1`).
> +- Three main US regions (`us-west1`,`us-east1`,`us-central1`).

Done

> ⬇️ Suggested change 
> -- 2 additional distant regions (`europe-west1`, `asia-northeast1`). Only [follower reads]({% link {{ page.version.version }}/follower-reads.md %}) will be made from these regions.
> +- Two additional distant regions (`europe-west1`, `asia-northeast1`). Only [follower reads]({% link {{ page.version.version }}/follower-reads.md %}) will be made from these regions.

Done

> ⬇️ Suggested change 
> -    To minimize the cross-region write latency, we will [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that the 3 US regions have voting replicas (that is, replicas that participate in [Raft quorum]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft)). This will leave the 2 additional non-US regions with only [non-voting replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#non-voting-replicas).
> +To minimize the cross-region write latency, [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that the 3 US regions have voting replicas (that is, replicas that participate in [Raft quorum]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft)). This will leave the 2 additional non-US regions with only [non-voting replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#non-voting-replicas).

Done

> ⬇️ Suggested change 
> -Next, enter the statements below at the SQL prompt. These statements will:
> +Next, enter the following statements at the SQL prompt. These statements will:

Done

> ⬇️ Suggested change 
> -- Make the [`movr` database]({% link {{ page.version.version }}/movr.md %}) into a [multiregion database]({% link {{ page.version.version }}/multiregion-overview.md %}) by [setting the primary region]({% link {{ page.version.version }}/alter-database.md %}#set-primary-region).
> +- Make the [`movr` database]({% link {{ page.version.version }}/movr.md %}) into a [multi-region database]({% link {{ page.version.version }}/multiregion-overview.md %}) by [setting the primary region]({% link {{ page.version.version }}/alter-database.md %}#set-primary-region).

Done.  But in protest!  e-mail became email eventually!!!!1!!!!

> ⬇️ Suggested change 
> -- Add the remaining regions we defined at cluster start time to the multi-region `movr` database (for a total of 5).
> +- Add the remaining regions defined at cluster start time to the multi-region `movr` database (for a total of 5).

Done

> Link ALTER LOCALITY.

Done

> ⬇️ Suggested change 
> -Now that our multiregion cluster is configured, we will use the `ALTER LOCALITY` statements shown below to [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that:
> +Now that the multi-region cluster is configured, use the `ALTER LOCALITY` statements shown below to [configure replication zones]({% link {{ page.version.version }}/configure-replication-zones.md %}) such that:

Done

> ⬇️ Suggested change 
> -- This will leave the 2 additional regions (Europe and Japan) with only [non-voting replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#non-voting-replicas) that can be used to serve [follower reads]({% link {{ page.version.version }}/follower-reads.md %}).
> +- The 2 additional regions (Europe and Japan) are left with only [non-voting replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#non-voting-replicas) that can be used to serve [follower reads]({% link {{ page.version.version }}/follower-reads.md %}).

Done

> In src/current/v23.2/cockroach-demo.md:

> ⬇️ Suggested change 
> -`--demo-locality` | <a name="demo-locality"></a> Specify [locality]({% link {{ page.version.version }}/cockroach-start.md %}#locality) information for each demo node. The input is a colon-separated list of key-value pairs, where the i<sup>th</sup> pair is the locality setting for the i<sup>th</sup> demo cockroach node.<br><br>For example, the following option assigns node 1's region to `us-east1` and availability zone to `1`, node 2's region to `us-east2` and availability zone to `2`, and node 3's region to `us-east3` and availability zone to `3`:<br><br>`--demo-locality=region=us-east1,az=1:region=us-east1,az=2:region=us-east1,az=3`<br><br>By default, `cockroach demo` uses sample region (`region`) and availability zone (`az`) replica localities for each node specified with the `--nodes` flag.
> +`--demo-locality` | <a name="demo-locality"></a> Specify [locality]({% link {{ page.version.version }}/cockroach-start.md %}#locality) information for each demo node. The input is a colon-separated list of key-value pairs, where the <i>i</i><sup>th</sup> pair is the locality setting for the <i>i</i><sup>th</sup> demo cockroach node.<br><br>For example, the following option assigns node 1's region to `us-east1` and availability zone to `1`, node 2's region to `us-east2` and availability zone to `2`, and node 3's region to `us-east3` and availability zone to `3`:<br><br>`--demo-locality=region=us-east1,az=1:region=us-east1,az=2:region=us-east1,az=3`<br><br>By default, `cockroach demo` uses sample region (`region`) and availability zone (`az`) replica localities for each node specified with the `--nodes` flag.

Done, but only because it looks better.  God markdown is a mess
